add new scheduler

JeanetteBruno opened the pull request:
This contains the same changes as pull request: added load based scheduling algorithm #489.  I'm creating this pull request in hopes that travis will behave better with a clean pull request. (Travis does not seem to be building for pr 489, I changed a travis integration setting after creating 489, so that may have some impact on how it is processing 489's changes.)

Commit:
added new scheduler

anthonyjpratti commented on common/stats/stats_names.go:
```diff
@@ -321,6 +321,47 @@ const (
+	/*
+		number of time Load Based Scheduler saw an unrecognized requestor
+	*/
+	SchedLBSUnknownJobCnt = "schedLBSUnknownJobCnt"
```
Why Cnt instead of Counter? Seems the convention is Counter.

anthonyjpratti commented:
Git commit titles should be in present tense.

"Add new scheduler"

Commit:
backing out stylistic changes

Commit:
updated as per review comments (from pr 489)

anthonyjpratti commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,780 @@
+	numWaitingTasks int // number of tasks still waiting to be started
+
+	tempEntitlement   int // temporary field to hold intermediate entitled num workers
+	tempNormalizedPct int // temporary field to hold the normalized load %
```
Please match above convention of comment before field.

anthonyjpratti commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,780 @@
+}
+
+// GetTasksToBeAssigned - the entry point to the load based scheduling algorithm
+// It returns the list of tasks that should be assigned to nodes as per the worker-class %s:
```
Consider rephrasing "assigned to nodes as per the worker-class %s:" as  "assigned to nodes according to worker-class"

anthonyjpratti commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,780 @@
+// It returns the list of tasks that should be assigned to nodes as per the worker-class %s:
+// Allocate available workers to classes based on target % allocations for each class and the current and
+// running load for each class.
+// - When a class has tasks waiting to start, the algorithm will determing the number of workers the class it 'entitled' to:
```
Grammatical error please rephrase

anthonyjpratti commented on perftests/scheduler_simulator/test_alg.go:
```diff
@@ -69,18 +69,20 @@ type timeSummary struct {
+	comparisonMap       map[string]*timeSummary
+	comparisonMapMu     sync.RWMutex
+	timeout             time.Duration
+	classLoadPercents       map[string]int32
```
Please format.

anthonyjpratti commented on perftests/scheduler_simulator/test_alg.go:
```diff
@@ -107,28 +109,26 @@ func MakeSchedulingAlgTester(testsStart, testsEnd time.Time, jobDefsMap map[int]
+		testsEnd:            testsEnd,
+		jobDefsMap:          jobDefsMap,
+		clusterSize:         clusterSize,
+		classLoadPercents:       classLoadPercents,
```
Please format.

anthonyjpratti commented on perftests/scheduler_simulator/test_alg.go:
```diff
@@ -166,32 +164,29 @@ func (st *SchedulingAlgTester) RunTest() error {
+			jobStart := simStart.Add(time.Duration(secondsAfterStart) * time.Second)
+			if time.Now().Before(jobStart) {
+				log.Warnf("now: %s, job start: %s", time.Now().Format(time.RFC3339), jobStart.Format(time.RFC3339))
+				<-time.After(jobStart.Sub(time.Now())) // wait till time reaches startTime0				<-
```
Why are you using time.After instead of time.Sleep?

JeanetteBruno replied:
I don't think it matters much... time. Sleep(d) will let the current goroutine enter sleeping sub-state, but still stay in running state, whereas, the channel receive operation <-time. After(d) will let the current goroutine enter blocking state.... I'm thinking that putting the test in blocking state might help with switching between threads.

anthonyjpratti commented on perftests/scheduler_simulator/test_alg.go:
```diff
@@ -166,32 +164,29 @@ func (st *SchedulingAlgTester) RunTest() error {
 			}
 
 			// give the job to the scheduler
+			log.Warnf("%s submitting job:jobType:%s, req:%s, tag:%s, basis: %s, tasks:%d", time.Now().Format(time.RFC3339), jobDef.JobType, jobDef.Requestor, jobDef.Tag, jobDef.Basis, len(jobDef.Tasks))
```
Does jobDef have a String method? Might make this clearer.

anthonyjpratti commented on perftests/scheduler_simulator/test_alg.go:
```diff
@@ -339,36 +323,37 @@ func (st *SchedulingAlgTester) getExternals(clusterSize int) *externalDeps {
+		waitingStatName := fmt.Sprintf("schedNumWaitingTasksGauge_%s", className)
+		runningCnt, ok1 := s[runningStatName]
+		waitingCnt, ok2 := s[waitingStatName]
+		if ok1 || ok2 {
```
Are you sure you want an or? Wouldn't that mean runningStatName or waitingStatName might be nil?

JeanetteBruno replied:
The intent of the condition is to exclude printing info on classes that don't have running or waiting tasks.

anthonyjpratti commented on scheduler/api/server.go:
```diff
@@ -78,3 +81,60 @@ func (h *Handler) GetSchedulerStatus() (*scoot.SchedulerStatus, error) {
+
+// GetRebalanceMinDuration get the duration(minutes) that the scheduler needs to be exceeding
+// the rebalance threshold before rebalancing.  <= 0 implies no rebalancing
+func (h *Handler) GetRebalanceMinDuration() (int32, error) {
```
This is ambiguous. Does min mean minimum or minute in this context?

JeanetteBruno replied:
changed all RebalanceMinDuration to RebalanceMinimumDuration

anthonyjpratti commented on scheduler/client/cli/sched_alg_params.go:
```diff
@@ -0,0 +1,174 @@
+type getLBSSchedAlgParams struct {
+	printAsJSON bool
+	params      lbsSchedAlgParams
+}
```
Why is this needed?

JeanetteBruno replied:
to allow us to query the scheduler for the current load based algorithm parameters.

anthonyjpratti commented on scheduler/client/cli/cli.go:
```diff
@@ -23,6 +24,18 @@ type simpleCLIClient struct {
 }
 
+// getReturnError extend the error with Invalid Request, Scoot server error, or Error getting status message
+func getReturnError(err error) error {
```
I think returnError is clear enough, and is the convention we use internally.

JeanetteBruno replied:
changed

anthonyjpratti commented on scheduler/client/cli/sched_alg_params.go:
```diff
@@ -0,0 +1,174 @@
+		fmt.Println("Rebalance Duration:", g.params.RebalanceMinDuration, " (minutes)")
+		log.Infof("Rebalance Threshold:%d", g.params.RebalanceThreshold)
+		fmt.Println("Rebalance Threshold:", g.params.RebalanceThreshold)
+
```
nit: no newline

anthonyjpratti commented on scheduler/client/cli/sched_alg_params.go:
```diff
@@ -0,0 +1,174 @@
+	return nil
+}
+
+func (s *lbsSchedAlgParams) registerFlags() *cobra.Command {
```
Why isn't this named setLbsSchedAlgParams

JeanetteBruno replied:
lbsSchedAlgParams is the struct that holds all the algorithm params.  It carries setter function, but is also used by the getter.  I could have created a setLBSAlgoParams that is strictly a pass through to the lbsSchedAlgParam struct to isolate the setter function but thought that approach was more complicated than necessary.

ryancouto replied:
I think I'd prefer that approach - isolating the setter

anthonyjpratti replied:
I'd follow what is convention in the other files, which appears to be a "set" struct and a "get" struct

JeanetteBruno replied:
I'm going to push back on this again.... my reasoning

1. this file makes 11 files in the directory (excluding doc.go) which moves us into 'create subdirectories range'
2. both cli subcommands use the lbsSchedAlgParams struct, so the getter would be using a data structure defined in the setter's go file adding a little more clumsiness to the code
3. it seem's very natural to me to have setter/getter functionality defined together.

but adding to mgci-1310 - who ever picks it up can make the change if they wish

anthonyjpratti replied:
Please follow the current conventions, it makes the code harder to read and reason about when deviations happen.

anthonyjpratti commented on scheduler/client/client.go:
```diff
@@ -35,10 +36,9 @@ func NewCloudScootClient(config CloudScootClientConfig) *CloudScootClient {
 func (c *CloudScootClient) RunJob(jobDef *scoot.JobDefinition) (r *scoot.JobId, err error) {
-	err = c.checkForClient()
-	if err != nil {
+	if err := c.checkForClient(); err != nil {
```
These stylistic changes are good changes, but do make it harder to focus on the important changes in this CL.

JeanetteBruno replied:
reverted

anthonyjpratti commented on scheduler/client/client.go:
```diff
@@ -166,6 +161,98 @@ func (c *CloudScootClient) GetSchedulerStatus() (*scoot.SchedulerStatus, error)
+	}
+	classLoadPercents, err := c.client.GetClassLoadPercents()
+	// if an error occurred reset the connection, could be a broken pipe or other
+	// unrecoverable error.  reset connection so a new clean one gets created
```
nit: two spaces are error should only be one

anthonyjpratti commented on scheduler/client/client.go:
```diff
@@ -166,6 +161,98 @@ func (c *CloudScootClient) GetSchedulerStatus() (*scoot.SchedulerStatus, error)
+		return nil, err
+	}
+	classLoadPercents, err := c.client.GetClassLoadPercents()
+	// if an error occurred reset the connection, could be a broken pipe or other
```
Why don't we do this for all methods?

JeanetteBruno replied:
good question - added it to the new functions

ryancouto replied:
Going further - why don't we just reset the connection regardless? Is there any value in persisting the client? Could be another cleanup item to add to MGCI-1310

JeanetteBruno replied:
added to ticket

anthonyjpratti commented on scheduler/client/scootcl/main.go:
```diff
@@ -17,6 +17,8 @@ import (
 //		get_scheduler_status
 //		set_scheduler_status [limit]
+//      get_scheduling_alg_params
+//      set_scheduling_alg_params
```
Please format. Looks like others used tabs over spaces.

JeanetteBruno replied:
done

anthonyjpratti commented on scheduler/server/job_state.go:
```diff
@@ -22,10 +26,13 @@ type jobState struct {
-	Completed  map[string]*taskState
-	Running    map[string]*taskState
-	NotStarted map[string]*taskState
+	// Completed and Running are only used by priority based scheduling algorithm
```
I saw pratios being removed. Are we not removing the priority based impl? If we don't remove that implementation, we should keep the tests associated with it, right?

JeanetteBruno replied:
fixed comment - these fields are being used

anthonyjpratti commented on scheduler/server/job_state.go:
```diff
@@ -42,6 +49,22 @@ type taskState struct {
+		return tasks
+	}
+
+	return nil
```
Map is a reference type, so these checks are not needed. If class is not in classTasksByStartTime, then classTasksByStartTime[class] will return nil, which means that classTasksByStartTime[class][start] will return nil. so this entire func can be a one liner.

JeanetteBruno replied:
data structure change eliminated the method

anthonyjpratti commented on scheduler/server/job_state.go:
```diff
@@ -170,12 +201,16 @@ func (j *jobState) taskCompleted(taskId string, running bool) {
 		delete(j.NotStarted, taskId)
```
That TODO on the prev line is worrying...

JeanetteBruno replied:
thanks for catching that.  I added it last year when I was starting to work with the (earlier) new algorithm and wasn't really sure if we needed to trap for error or what..... since then I realized we don't need it.

anthonyjpratti commented on scheduler/server/job_state.go:
```diff
@@ -197,3 +234,43 @@ func (j *jobState) getJobStatus() domain.Status {
+func (j *jobState) addTaskToStartTimeMap(jobClass string, task *taskState, startTimeSec time.Time) {
+	if j.tasksByJobClassAndStartTimeSec == nil {
+		return
+	}
```
Is this an ok state for this to be in?

JeanetteBruno replied:
added logging an error

anthonyjpratti commented on scheduler/server/job_state.go:
```diff
@@ -197,3 +234,43 @@ func (j *jobState) getJobStatus() domain.Status {
+	if _, ok := j.tasksByJobClassAndStartTimeSec[jobClass][startTimeSec]; !ok {
+		j.tasksByJobClassAndStartTimeSec[jobClass][startTimeSec] = taskStateByTaskID{}
+	}
+	j.tasksByJobClassAndStartTimeSec[jobClass][startTimeSec][fmt.Sprintf("%s_%s", task.JobId, task.TaskId)] = task
```
Keys can be structs, please prefer that over concatenation.

JeanetteBruno replied:
done

anthonyjpratti commented on scheduler/server/job_state.go:
```diff
@@ -197,3 +234,43 @@ func (j *jobState) getJobStatus() domain.Status {
+	if _, ok := j.tasksByJobClassAndStartTimeSec[jobClass]; !ok {
+		j.tasksByJobClassAndStartTimeSec[jobClass] = taskTimeBuckets{}
+	}
+	if _, ok := j.tasksByJobClassAndStartTimeSec[jobClass][startTimeSec]; !ok {
```
Looking up on jobClass is unnecessary. if we do an early return when ok is true, we can then preserve the reference to both taskTimeBuckets and taskStateByTaskID.

JeanetteBruno replied:
done

anthonyjpratti commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,780 @@
+var DefaultLoadBasedSchedulerClassPercents = map[string]int32{"c0": 100}
+var DefaultRequestorToClassMap = map[string]string{".*": "c0"}
+var DefaultMinRebalanceTime = time.Duration(4 * time.Minute)
+var MaxTaskDuration = time.Duration(4 * time.Hour)
```
Prefer var blocks

JeanetteBruno replied:
done

anthonyjpratti commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,780 @@
+
+type LoadBasedAlgConfig struct {
+	classLoadPercents           map[string]int
+	classLoadPercentsMu         sync.RWMutex
```
Please format.

anthonyjpratti commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,780 @@
+
+	// local copy of load pcts and requestor map to use during assignment computation
+	// to insulate the computation from external changes to the configuration
+	classLoadPercents         map[string]int
```
Please format.

anthonyjpratti commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,780 @@
+	stat stats.StatsReceiver
+}
+
+// LoadBasedAlg the scheduling algorithm allocates job tasks to workers using a class map as
```
In the README, you say that workers are allocated to tasks and it is flipped here. So which is it?

JeanetteBruno replied:
rewrote comments/README

anthonyjpratti commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,780 @@
+// follows:
+// - each job maps to a 'class' (based on the job's requestor value)
+// - classes are assigned a % of the number of scoot workers.
+// When the algorithm is assigning tasks to workers it will try to start tasks such that the
```
super nit: missing comma after introductory clause.

JeanetteBruno replied:
rewrote the whole section.

anthonyjpratti commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,780 @@
+
+	rebalanced := false
+	var stopTasks []*taskState
+	if lbs.getRebalanceMinDuration() > 0 && lbs.getRebalanceThreshold() > 0 {
```
Why not call getCurrentPercentsSpread here? If it's a perf concern, could you please comment that here.

anthonyjpratti commented on scheduler/server/orig_scheduling_alg.go:
```diff
@@ -7,10 +7,42 @@ import (
 
-type OrigSchedulingAlg struct{}
+// Decrease the NodeScaleFactor by taking the percentage defined by NodeScaleAdjust[Priority].
+// Note: the use case here is to hit an SLA for each job priority, and this is a coarse way to do so.
```
We have SLAs for scoot?

anthonyjpratti commented on scheduler/server/orig_scheduling_alg.go:
```diff
@@ -7,10 +7,42 @@ import (
 	"github.com/twitter/scoot/scheduler/domain"
 )
 
-type OrigSchedulingAlg struct{}
```
I'm surprised this files contents changed. Shouldn't it stay the same mostly if the impl hasn't changed?

anthonyjpratti commented on scheduler/server/orig_scheduling_alg.go:
```diff
@@ -7,10 +7,42 @@ import (
+		totalOutstandingTasks += (len(j.Tasks) - j.TasksCompleted)
+	}
+
+	// Udate SoftMaxSchedulableTasks based on number of healthy nodes and the total number of tasks.
```
nit: typo Udate to Update

JeanetteBruno replied:
removed file

anthonyjpratti commented on scheduler/server/orig_scheduling_alg_test.go:
```diff
@@ -23,9 +308,9 @@ func Test_StatefulSchedulerRequestorCounts(t *testing.T) {
-	time.Sleep(50 * time.Millisecond) // let task threads run to complete
-	s.step()                          // get first set of completed tasks and start next set
+	s.step()                           // start the first set of tasks
+	time.Sleep(500 * time.Millisecond) // let task threads run to complete
```
This seems like a pretty large increase. Do you know why that is needed?

JeanetteBruno replied:
removed file

anthonyjpratti commented on scheduler/server/job_state.go:
```diff
@@ -22,10 +26,13 @@ type jobState struct {
-	Completed  map[string]*taskState
-	Running    map[string]*taskState
-	NotStarted map[string]*taskState
+	// Completed and Running are only used by priority based scheduling algorithm
```
We deleted the priority algo below, so why do we have these fields still?

JeanetteBruno replied:
they are used by the new scheduling algorithm - I'll update the comment.

anthonyjpratti commented on scheduler/server/priority_based_scheduling_alg.go:
```diff
@@ -1,255 +0,0 @@
-package server
```
If this is deleted, why do we still have comments that reference the priority based algo?

JeanetteBruno replied:
fixed comments

anthonyjpratti commented on scheduler/server/stateful_scheduler.go:
```diff
@@ -267,12 +259,22 @@ func NewStatefulScheduler(
-		config.SoftMaxSchedulableTasks = DefaultSoftMaxSchedulableTasks
+	config.SchedAlgConfig = &LoadBasedAlgConfig{
+		rebalanceThreshold:   0,
+		rebalanceMinDuration: 0 * time.Minute,
```
Zero values here are already both zero, so this is not needed.

JeanetteBruno replied:
removed

JeanetteBruno commented on scheduler/server/task_scheduler.go:
```diff
@@ -20,101 +20,108 @@ type taskAssignment struct {
-		cfg = *config
+	tasks, stopTasks := s.config.SchedAlg.GetTasksToBeAssigned(jobs, stat, cs, requestors)
+	// Exit if no tasks qualify to be scheduled.
+	if len(tasks) == 0 {
```
if len(stopTasks) > 0 then log an error - this should never happen

also verify that task is aborted (blocking) before moving on to starting tasks.

JeanetteBruno replied:
abort is blocking, added error message

JeanetteBruno commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,780 @@
+		var jc *jobClass
+		var ok bool
+		className := GetRequestorClass(requestor, lbs.requestorReToClassMap)
+		if className != "" {
```
don't need to test for className != "" , !ok should be sufficient to know to use SchedLBSUnknownJobCounter and classNameWithLeastWorkers

JeanetteBruno commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,780 @@
+		}
+	}
+
+	return maxPct - minPct
```
should this be subtract or add (minPct will be < 0 for loaned workers) - check the spreadsheet calc.

JeanetteBruno replied:
looked at it some more - think its ok (for now - till we have a better handle on thresholds, etc)

Commit:
updates from review comments

dgdonut commented on perftests/scheduler_simulator/fake_worker_cli.go:
```diff
@@ -76,7 +76,9 @@ func (fw *FakeWorker) Run(cmd *runner.Command) (runner.RunStatus, error) {
 	fw.state = runner.RUNNING
 	go func(fw *FakeWorker) {
-		time.Sleep(time.Duration(duration) * time.Second)
+		t := time.NewTicker(time.Duration(duration) * time.Second)
```
<-time.After(...) would do the same thing

JeanetteBruno replied:
I know - I happened to read that  using ticker/stop is more efficient wrt to garbage collection when I was looking at this part of the code so converted it. I expect we'll increase the simulation to use 10K workers managing 100k tasks as some point, so converted it.

ryancouto replied:
could also just skip assigning to t
`<-time.NewTicker(time.Second).C`

anthonyjpratti replied:
Is there a reason to prefer <-time.After over Sleep?

JeanetteBruno replied:
need access to t so we can stop the ticker.

JeanetteBruno replied:
> Is there a reason to prefer <-time.After over Sleep?

see comment above- ticker is more efficient wrt to garbage collection.

dgdonut commented on scheduler/server/job_state.go:
```diff
@@ -151,13 +171,18 @@ func (j *jobState) taskStarted(taskId string, tr *taskRunner) {
 	j.Running[taskId] = taskState
+
+	// add the task to the map of tasks by start time
+	startTimeSec := taskState.TimeStarted.Truncate(time.Second)
```
We are always truncating to the nearest time.Second but this is not enforced anywhere (e.g. by some sort of wrapping mechanism). If we ever change or add code and forget to do this it will probably break things.

JeanetteBruno replied:
it's only local to this function - the whole intent of this function is to step backwards in time 1 second at a time.  Only the initializer (ln 572) and time decrement (ln 584)  should be modifying the time value.  If we're doing something else to modify the time value, yes we would need to compute the time then 'wrap' with Truncate() (In essence Truncate() is the only wrapper we could have in common to all lines of code impacting the time index.)

dgdonut commented on scheduler/domain/definitions.go:
```diff
@@ -52,6 +52,10 @@ type JobDefinition struct {
+func (jd *JobDefinition) String() string {
+	return fmt.Sprintf("jobType:%s, req:%s, tag:%s, basis: %s, tasks:%d", jd.JobType, jd.Requestor, jd.Tag, jd.Basis, len(jd.Tasks))
+}
+
```
Somewhat of a tangent - what is happening with the Priority constructs defined in this file and used in various places?

JeanetteBruno replied:
It's not used anymore, but I'm not taking it out due to the extent of the changes.  
I've created https://jira.twitter.biz/browse/MGCI-1310 for clean up issues.

dgdonut commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,781 @@
+
+// defaults for the LoadBasedScheduler algorithm: only one class and all jobs map to that class
+var (
+	DefaultLoadBasedSchedulerClassPercents = map[string]int32{"c0": 100}
```
Realizing now, we are making the scheduler more stateful. When schedulers restart they will lose any configured class %s. We will need a mechanism to preserve this between instances.

JeanetteBruno replied:
we need to talk about this, I think there are other parameters (throttle?) we might also want to preserve... think for now we should set meaningful default values (after running simulations to understand what they should be).  Then have another ticket for preserving parameter values across restarts.

dgdonut commented on scheduler/server/stateful_scheduler.go:
```diff
@@ -1162,6 +1146,29 @@ func (s *statefulScheduler) processKillJobRequests(reqs []jobKillRequest) {
+		task.TaskRunner.Abort(true, UserRequestedErrStr)
+	} else if task.Status == domain.NotStarted {
+		st := runner.AbortStatus("", tags.LogTags{JobID: jobState.Job.Id, TaskID: task.TaskId})
+		st.Error = UserRequestedErrStr
```
It's not UserRequested

JeanetteBruno replied:
thanks! made into parameter.  Rebalance aborts will use "RebalanceRequested"

dgdonut commented on scheduler/server/stateful_scheduler.go:
```diff
@@ -1162,6 +1146,29 @@ func (s *statefulScheduler) processKillJobRequests(reqs []jobKillRequest) {
 	}
 }
 
+// abortTask abort a task - will be triggered by killing a job and when the scheduling algorithm rebalances the workers
```
Is this substantially different than the existing abort task mechanisms? Why did we need it?

JeanetteBruno replied:
no - it's copying the abort functionality from inline in processKillJobRequests() and exposing it so it can be used by both processKillJobRequests(), and the taskScheduler.go getTaskAssignments when stopping tasks for rebalancing.

dgdonut commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,781 @@
+	tasksByJobClassAndStartTimeSec map[taskClassAndStartKey]taskStateByJobIDTaskID
+}
+
+// NewLoadBasedAlg allocate a new LoadBaseSchedAlg object.  If the load %'s don't add up to 100
```
Doesn't seem to return errors or care what the %s add up to

JeanetteBruno replied:
it normalizes the %s automatically

dgdonut replied:
I mean - the function is not doing what the comment is saying

dgdonut reviewed (commented):
How is default class bucketing going to work, I don't see this mentioned anywhere any more.

Commit:
updated as per review comments and fixed simulator bugs and logging

ryancouto commented on scheduler/client/cli/sched_alg_params.go:
```diff
@@ -0,0 +1,174 @@
+	return r
+}
+
+func (s *lbsSchedAlgParams) run(cl *simpleCLIClient, cmd *cobra.Command, args []string) error {
```
I understand the desire to minimize the # of files, but I personally would prefer to have separate files for each command.

JeanetteBruno replied:
see comment above

ryancouto commented on scheduler/server/job_state.go:
```diff
@@ -21,11 +24,14 @@ type jobState struct {
+	// Completed and Running are only used by the scheduling algorithm
+	Completed  taskStateByTaskID
+	Running    taskStateByTaskID
+	NotStarted taskStateByTaskID
```
I think we should consider renaming NotStarted to Waiting. We use both interchangeably, but Waiting seems like a better descriptor imo.

JeanetteBruno replied:
added to mgci-1310

ryancouto commented on scheduler/server/job_state.go:
```diff
@@ -42,6 +48,17 @@ type taskState struct {
 type taskStatesByDuration []*taskState
 
+// the following types are used to access task State objects by class, startTime, taskID
+type taskClassAndStartKey struct {
```
don't think we need to include "key" in this name. the fact that it's used as a key in maps doesn't tell me anything about the struct itself.

JeanetteBruno replied:
I have 'Key' in it as an indicator that it is only used (useful) for access the map (allowing us to find tasks by class and start time)

ryancouto commented on scheduler/server/job_state.go:
```diff
@@ -42,6 +48,17 @@ type taskState struct {
+	class string
+	start time.Time
+}
+type jobIDTaskIDKey struct {
```
ditto re: "key"

JeanetteBruno replied:
see comment above

ryancouto commented on scheduler/server/job_state.go:
```diff
@@ -197,3 +228,39 @@ func (j *jobState) getJobStatus() domain.Status {
+		log.Errorf("tasksByJobClassAndStartTime map not found.  Skipping adding task to it. jobID: %s, taskID:%s, jobClass:%s, startTime:%s", task.JobId, task.TaskId, jobClass, startTimeSec.Format("2006-01-02 15:04:05 -0700 MST"))
+		return
+	}
+	timeBucket := taskClassAndStartKey{class: jobClass, start: startTimeSec}
```
it's a timeBucket, but it's also a classBucket, right?

JeanetteBruno replied:
actually it's the key for a set of tasks bucketed by class/time... changed

ryancouto commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,781 @@
+	maxTaskRunningMapIndex int
+
+	origNumWaitingTasks int
+	origNumRunningTasks int
```
let's add comments for these fields too

ryancouto commented on scheduler/server/task_scheduler.go:
```diff
@@ -13,108 +13,121 @@ type taskAssignment struct {
+	cs := s.clusterState
+	jobs := s.inProgressJobs
+	requestors := s.requestorMap
+	stat := s.stat
```
why make these assignments?

JeanetteBruno replied:
was trying to minimize code changes from moving the function into stateful scheduler, but then ended up having to change the whole function anyway, so will remove.
Thanks for drawing my attention to it.

ryancouto commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,781 @@
+	log.Debugf("in LoadBasedAlg.GetTasksToBeAssigned: numWorkers:%d, numIdleWorkers:%d", len(cs.nodes), cs.numFree())
+
+	// make local copies of the load pct structures
+	lbs.classLoadPercents = lbs.LocalCopyClassLoadPercents()
+	lbs.requestorReToClassMap = lbs.getRequestorToClassMap()
+	lbs.classByDescLoadPct = lbs.getClassByDescLoadPct()
```
why do we assign these here if we already have access to it via the config?

ryancouto commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,781 @@
+	rebalanced := false
+	var stopTasks []*taskState
+	if lbs.getRebalanceMinimumDuration() > 0 && lbs.getRebalanceThreshold() > 0 {
+		// currentPctSpread is the delta between the highest and lowest
```
between highest and lowest what?

Commit:
downgraded error to warning

Commit:
Merge branch 'master' into jbruno/SchedAlgWithEntitlementPcts

JeanetteBruno commented:
> How is default class bucketing going to work, I don't see this mentioned anywhere any more.
not sure if your asking about unkown requestors or default class percents and requestor to class maps so I'll answer both....

**default class %s and requestor map (default values on scheduler restart):**
stateful_scheduler.go creates the LoadBasedAlg giving it DefaultLoadBasedSchedulerClassPercents, and DefaultRequestorToClassMap.  We'll initialize those 2 constants with the values determined by our simulation runs.

**unknown requestors:**
the default class %s will include an unknown requestor class that is allocated 1% of the workers.  jobs with unknown requestors will be assigned to this class.

ryancouto commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,781 @@
+		currentPctSpread := lbs.getCurrentPercentsSpread(numWorkers)
+		if currentPctSpread > lbs.getRebalanceThreshold() {
+			nilTime := time.Time{}
+			if lbs.exceededRebalanceThresholdStart == nilTime {
```
when would this be true? just on the first iteration?

JeanetteBruno replied:
this is true when a lot of long running tasks have been started on 'loaned' workers.

ryancouto commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,781 @@
+	}
+
+	log.Debugf("Returning %d start tasks, %d stop tasks", len(tasksToStart), len(stopTasks))
+	return tasksToStart, stopTasks
```
let's stay consistent with naming - tasksToStart & tasksToStop or startTasks & stopTasks

ryancouto commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,781 @@
+	tasks := []*taskState{}
+
+	startingTaskCnt := 0
+	// work our way through the class's jobs, starting with jobs with the least number of running tasks,
```
why least running tasks vs lowest pct of running tasks relative to waiting?

JeanetteBruno replied:
lowest pct of running tasks would allow jobs with many tasks to starve jobs with few tasks.  I think trying to have an equal number of tasks running for each job in a class is a better solution.

ryancouto commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,781 @@
+				// get the next task to start from the job
+				tasks = append(tasks, lbs.getJobsNextTask(job))
+
+				// move the job to jobsByRunning tasks with numRunningTasks + 1 entry.  Note: we don't have to pull it from
```
*jobsByRunningTasks

ryancouto commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,781 @@
+		}
+	}
+
+	return tasks // note: we should never hit this line
```
why not? why include it?

ryancouto commented on scheduler/server/load_based_sched_alg.go:
```diff
@@ -0,0 +1,781 @@
+		pctTotal += int(val)
+	}
+	if pctTotal != 100 {
+		log.Errorf("LoadBalanced scheduling %%'s don't add up to 100, normalizing them")
```
let's break normalization out into a helper so we can more easily test it

JeanetteBruno replied:
done

Commit:
updates as per review comments

Commit:
add prod classes and load %s, plus some readability changes

Commit:
extract percents normalize to its own function, add test for normalizing percents, plus misc style change

Commit:
Merge branch 'master' into jbruno/SchedAlgWithEntitlementPcts

JeanetteBruno commented:
wrt to Ryan's question: _when would this be true? just on the first iteration?_ (when is currentPctSpread > lbs.getRebalanceThreshold()) - this is true when a lot of long running tasks have been started on loaned workers. (see algorithm summary above LoadBasedAlg struct definition and in README.md)

wrt to Ryan's question: _why least running tasks vs lowest pct of running tasks relative to waiting?_ we could choose to change the algorithm to that, imo it is more fair to try to give all jobs within a class an equal number of workers.  If we go with lowest pct of running to waiting tasks we could end up with jobs with lots of tasks starving jobs with small number of tasks.

wrt to Ryan's question: _why not? why include it?_. I added a log statement and updated the comment... the algorithm should always return in the loop above, but needed to include a return statement outside of the loop for the code to compile.

Commit:
comment/variable name updates and delete dead doc.go

Commit:
reorg scheduler alg cli getter/setter, update default class percent definitions and requestor map

willnorris commented:
closing PR (and will immediately reopen) to try and bump CLA Assistant.
